FutureNHS/futurenhs-platform
# Create 'get workspace permissions' query and frontend Error component.

![Screenshot 2020-11-11 at 09 00 23](https://user-images.githubusercontent.com/22911872/98795542-3b65e480-2402-11eb-86bf-55461e5eeb1e.png)

- Ticket: [AB#1818](https://dev.azure.com/futurenhs/75407963-f812-43e5-a734-1059bbc036a3/_workitems/edit/1818)

## Acceptance criteria

- Workspace Service has a query that informs us of whether or not a given user is a member of a workspace.

- The frontend can render an error component when a user doesn't have sufficient permissions to access or edit a folder.

## Pre-review checklist

- [x] Analytics (user analytics, e.g. Google Analytics)
- [x] Observability (metrics/tracing)
- [x] Data columns adhere to the [Dublin Core Metatdata schema](https://www.dublincore.org/specifications/dublin-core/dcmi-terms/). Further information is available [here](https://www.gov.uk/government/publications/recommended-open-standards-for-government/using-metadata-to-describe-data-shared-within-government)

## Testing information

- Test plan?

1. Change the permissions on a folder to be 'workspace members', by navigating to it's edit page and selecting 'Make visible to all members of this workspace' from the permissions section. Save the changes. grab the folder URL for later. 
2. Using tablePlus or similar, access the local DB.
3. Grab the ID of a user who belongs to the workspace you've just updated a folder in.
4. Run the local workspace service with `make run-local` and navigate to http://localhost:3030/graphiql.
5. Run the below query, using the workspace ID from the folder URL, and the user ID of the user from 2 steps above. In addition, add the following to the headers section in the bottom right `{ "x-user-auth-id": "feedface-0000-0000-0000-000000000000"}`.

```graphql
mutation {
  changeWorkspaceMembership(
    input: {
      workspace: [workspace id here]
      user: [user id here]
      newRole: NON_ADMIN
    }
  ) {
    id
    title
    admins: members(filter: ADMIN) {name, id}
    members: members(filter: NON_ADMIN) {name, id}
  }
}
```

6. In the frontend, logout with `localhost:3000/auth/logout`.
7. In the file `noop-passport-strategy.js`, change lines 15-17 to reflect the details of the user who's ID you used in the above query. 
8. Log back in with `localhost:3000/auth/login`. You are now authenticated with the user that has insufficient access permissions for the folder edited earlier.
9. Navigate to the folder URL from step 1. You should see an error.

## Test checklist


- Platforms/Browsers:

  - [x] Google Chrome (latest version)

- Accessibility:

  - [x] Magnification
  - [x] Text size 200%
  - [x] Touch target areas (Mobile)
  - [x] Landscape (Mobile)

![](https://media.giphy.com/media/xT8qBcsMQwZTvkxFnO/giphy.gif)
